pkg/workloads/console/runner/runner.go

```diff
@@ -211,6 +228,56 @@ func (c *Runner) WaitUntilReady(ctx context.Context, createdCsl workloadsv1alpha
 	}
 }
 
+func (c *Runner) waitForRoleBinding(ctx context.Context, csl *workloadsv1alpha1.Console) error {
+	rbClient := c.kubeClient.RbacV1().RoleBindings(csl.Namespace)
```

Comment:
Does this become much simpler both to implement and test if we poll?

```go
for {
	select {
	case <-ctx.Done():
		return errors.New("timed out waiting for rolebinding to be created")
	default:
		if rb, err := rbClient.Get(csl.Name, metav1.GetOptions{}); err == nil {
			break // found
		}

		time.Sleep(interval)
	}
}
```

Reply:
It does, but the using a watcher felt like a more "kube-native" (™️) way of doing this, and there was already analogous code right next to this using a console watcher. You still need to perform a Get to mitigate the race with opening the watcher. I'll remove all watcher code from both polling loops and we can see how that looks.

Reply:
Yeah I hadn't properly looked at the rest of the file. I now see we're repeating this pattern everywhere.

Two things:

- I think this is the proper way of doing this
- I think the way we do it right now complicates our code

The second part wouldn't be half as bad if it didn't commit us to making any stubbing tests quite nasty, and introducing lots of async pain to our testing setup. We can solve this by either using more realistic tests (really interacting with kubernetes) or by creating a better concept of "watch this thing, but also inject the current value if it's already there".

Given some tests will always be unit tests, in the past I've gone with creating an abstraction I mentioned above, the "get and watch" or however you term it. You then properly test that code and can trust the implementation to do the right thing elsewhere, and for unit tests replace it with a reliable mock.

This is great and all, but I didn't realise I was opening pandoras box here. I think we might be better off going with this implementation now if it works and then ticketing this up as a possible improvement later. Perhaps even just opening an issue in the github repo about this as an optimisation. I'm keen to merge something that works for this now to enable console rollout and revisit the implementation later.

Wdyt?

Reply:
Totally agree! Let's merge this now as it works and I'll PR that improvement next. Small steps and all.